### PR TITLE
Flat JSON for pubsub and text files

### PR DESCRIPTION
**Pull request by tims**
kafka will be a follow up PR

### CODE REVIEW

File: ingestion/src/main/java/feast/source/common/ParseJsonTransform.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+      }
+      return gson.fromJson(json, valueMapType);
+    }
+  }
+
+  private static class ValueDeserializer implements JsonDeserializer<Value> {
+
+    @Override
+    public Value deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
+        throws JsonParseException {
```
Comment (woop) [inline ingestion/src/main/java/feast/source/common/ParseJsonTransform.java]: Can this possibly be flattened a bit with [guard clauses](https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html)?

File: ingestion/src/main/java/feast/source/common/ValueMapToFeatureRowTransform.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+          Value value = entry.getValue();
+          Field field = fields.get(name);
+
+          if (value == null || value.getValCase().equals(ValCase.VAL_NOT_SET)) {
+            continue;
+          }
+
+          // A feature can only be one of these things
+          if (entityIdColumn.equals(name)) {
+            builder.setEntityKey(Values.asString(value).getStringVal());
```
Comment (woop) [inline ingestion/src/main/java/feast/source/common/ValueMapToFeatureRowTransform.java]: What happens if there are multiple columns with the same name?

File: ingestion/src/main/java/feast/source/pubsub/PubSubFeatureSource.java [hunk 3]
```diff
@@ -54,53 +60,94 @@
...
+            .apply(new StringToValueMapTransform())
+            .apply(new ValueMapToFeatureRowTransform(importSpec.getEntities(0),
+                importSpec.getSchema()));
+        break;
+      case JSON:
+        Preconditions.checkArgument(importSpec.getEntitiesCount() == 1,
+            "pubsub source with format json, import spec must have one entity");
+        featureRows = input.getPipeline()
+            .apply(fromSubscriptionOrTopic(PubsubIO.readStrings(), options))
+            .apply(ParseCsvTransform.builder().header(Lists.newArrayList()).build())
```
Comment (woop) [inline ingestion/src/main/java/feast/source/pubsub/PubSubFeatureSource.java]: Should this be JSON?

File: ingestion/src/main/java/feast/source/common/ValueMapToFeatureRowTransform.java (hunk 2 above)
Comment (tims) [reply ingestion/src/main/java/feast/source/common/ValueMapToFeatureRowTransform.java]: This is impossible if it's from a bq or json source. For csv, you'd have to specify multiple fields with the same name in the import spec, and that causes an IllegalArgumentException in the csv reader aborting the job.

File: ingestion/src/main/java/feast/source/common/ParseJsonTransform.java (hunk 1 above)
Comment (tims) [reply ingestion/src/main/java/feast/source/common/ParseJsonTransform.java]: We decided we are conforming with the google java style guide, so we always need braces even when they are optional.
https://google.github.io/styleguide/javaguide.html#s4.1.1-braces-always-used

File: ingestion/src/main/java/feast/source/pubsub/PubSubFeatureSource.java (hunk 3 above)
Comment (tims) [reply ingestion/src/main/java/feast/source/pubsub/PubSubFeatureSource.java]: thanks! fixed

File: ingestion/src/main/java/feast/source/common/ParseJsonTransform.java (hunk 1 above)
Comment (woop) [reply ingestion/src/main/java/feast/source/common/ParseJsonTransform.java]: Guard clauses are different to optional braces. It's more about flattening nested if-else statements for readability.

It's really up to the developer which they prefer, so won't nitpick.

**Pull request merged**